distributed/client.py

```diff
@@ -4906,8 +4906,22 @@ def register_worker_callbacks(self, setup=None):
 
     async def _register_worker_plugin(self, plugin=None, name=None, nanny=None):
         if nanny or nanny is None and isinstance(plugin, NannyPlugin):
+            if not isinstance(plugin, NannyPlugin):
+                warnings.warn(
+                    "Registering duck-typed plugins has been deprecated. "
+                    "Please make sure your plugin subclasses `NannyPlugin`.",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
             method = self.scheduler.register_nanny_plugin
         else:
+            if not isinstance(plugin, WorkerPlugin):
+                warnings.warn(
+                    "Registering duck-typed plugins has been deprecated. "
+                    "Please make sure your plugin subclasses `WorkerPlugin`.",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
```

Comment:
Maybe it would become more readable like this?

```python
if isinstance(plugin, NannyPlugin):
    method = self.scheduler.register_nanny_plugin
elif isinstance(plugin, WorkerPlugin):
    method = self.scheduler.register_worker_plugin
elif nanny:
   warnings.warn(...)
   method = self.scheduler.register_nanny_plugin
else:
    warnings.warn(...)
    method = self.scheduler.register_worker_plugin
```

Reply:
I would like to leave this functionally the same for this PR and deprecate `register_{worker|scheduler}_plugin` in favor of a dispatched `register_plugin` in a follow-up PR.